sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzurePath.java

```diff
@@ -58,162 +120,325 @@ public Path getRoot() {
-        return null;
+        if (this.withoutRoot().isEmpty()) {
+            return null;
+        } else {
+            return this.parentFileSystem.getPath(Flux.fromArray(this.splitToElements()).last().block());
+        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public Path getParent() {
-        return null;
+        /*
+        If this path only has one element, there is no parent. Note the root is included in the parent, so we don't
+        use getNameCount here.
+         */
+        if (this.splitToElements().length == 1) {
+            return null;
+        }
+
+        return this.parentFileSystem.getPath(
+            this.pathString.substring(0, this.pathString.lastIndexOf(this.parentFileSystem.getSeparator())));
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public int getNameCount() {
-        return 0;
+        return this.splitToElements(this.withoutRoot()).length;
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public Path getName(int i) {
-        return null;
+        if (i < 0 || i >= this.getNameCount()) {
+            throw new IllegalArgumentException();
+        }
+        return this.parentFileSystem.getPath(this.splitToElements(this.withoutRoot())[i]);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public Path subpath(int i, int i1) {
-        return null;
+    public Path subpath(int begin, int end) {
+        if (begin < 0 || begin >= this.getNameCount()
+            || end <= begin || end > this.getNameCount()) {
+            throw new IllegalArgumentException(String.format("Values of begin: %d and end: %d are invalid",
+                begin, end));
+        }
+
+        Iterable<String> subnames = Flux.fromArray(this.splitToElements(this.withoutRoot()))
+            .skip(begin)
+            .take(end - begin)
+            .toIterable();
+
+        return this.parentFileSystem.getPath(String.join(this.parentFileSystem.getSeparator(), subnames));
     }
 
     /**
+     * In this implementation, a root component starts with another root component if the two root components are
+     * equivalent strings. In other words, if the files are stored in the same container.
+     *
      * {@inheritDoc}
      */
     @Override
     public boolean startsWith(Path path) {
-        return false;
+        /*
+        There can only be one instance of a file system with a given id, so comparing object identity is equivalent
+        to checking ids here.
+         */
+        if (path.getFileSystem() != this.parentFileSystem) {
+            return false;
+        }
+
+        String[] thisPathElements = this.splitToElements();
+        String[] otherPathElements = ((AzurePath) path).splitToElements();
+        if (otherPathElements.length > thisPathElements.length) {
+            return false;
+        }
+        for (int i = 0; i < otherPathElements.length; i++) {
+            if (!otherPathElements[i].equals(thisPathElements[i])) {
```

Comment:
What about a path with ".."? In other words, should we normalize first?

Reply:
I think it makes sense to normalize before comparing

Reply:
It looks like the default system (which I'm generally trying to follow) doesn't normalize and will return false for "foo/bar" starts with "foo/.", so I'll leave this as is.